File: go/xrr.go
package xrr

import (
	"context"
	"errors"
)

// Mode controls session behaviour.
type Mode string

const (
	ModeRecord      Mode = "record"
	ModeReplay      Mode = "replay"
	ModePassthrough Mode = "passthrough"
)

// ErrCassetteMiss is returned when replay finds no matching cassette file.
var ErrCassetteMiss = errors.New("xrr: cassette miss")

// Adapter intercepts one channel type.
type Adapter interface {
	ID() string
	Fingerprint(req Request) (string, error)
	Serialize(v any) ([]byte, error)
	Deserialize(data []byte, target any) error
}

// Request is an opaque adapter-defined value.
type Request interface {
	AdapterID() string
}

// Response is an opaque adapter-defined value.
type Response interface {
	AdapterID() string
}

// Cassette reads/writes interaction files.
//
// Save persists a recorded interaction. recordedErr is the error returned by
// the original do() (or nil); when non-nil it is serialized as the envelope
// error field on the resp file so replay can re-emit a matching error.
//
// Load returns recordedErr as a string from the resp envelope's optional
// error field. Empty string ⇒ success; non-empty ⇒ replay should surface
// errors.New(recordedErr) alongside respTarget. ErrCassetteMiss is returned
// if no matching files exist.
type Cassette interface {
	Load(adapterID, fingerprint string, reqTarget, respTarget any) (recordedErr string, err error)
	Save(adapterID, fingerprint string, req, resp any, recordedErr error) error
}

// Session owns the lifecycle of one record/replay run.
type Session interface {
	Record(ctx context.Context, adapter Adapter, req Request,
		do func() (Response, error)) (Response, error)
	Close() error
}

File: ts/src/xrr.ts
/**
 * Core xrr interfaces — mirrors Go reference implementation.
 */

export type Mode = "record" | "replay" | "passthrough";

export const ErrCassetteMiss = new Error("xrr: cassette miss");

export interface Adapter<Req, Resp> {
  id: string;
  fingerprint(req: Req): Promise<string>;
  serializeReq(req: Req): unknown;
  serializeResp(resp: Resp): unknown;
  deserializeReq(data: unknown): Req;
  deserializeResp(data: unknown): Resp;
}

export interface Cassette {
  save(adapterID: string, fingerprint: string, req: unknown, resp: unknown): Promise<void>;
  load(adapterID: string, fingerprint: string): Promise<{ req: unknown; resp: unknown }>;
}

export interface Session {
  record<Req, Resp>(
    adapter: Adapter<Req, Resp>,
    req: Req,
    do_: () => Promise<Resp>
  ): Promise<Resp>;
}
